src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -39,52 +49,50 @@ class LinearIsotropicConductor {
                        "Specific heat capacity must be positive in the linear isotropic conductor material model.");
   }
 
-  /**
-   * @brief Function defining the thermal flux (constitutive response)
-   *
-   * @tparam T1 type of the temperature (e.g. tensor or dual type)
-   * @tparam T2 type of the temperature gradient (e.g. tensor or dual type)
-   * @param temperature_gradient Gradient of the temperature (du_dx)
-   * @return The thermal flux of the material model
-   */
-  template <typename T1, typename T2>
-  SERAC_HOST_DEVICE T2 operator()(const T1& /* temperature */, const T2& temperature_gradient) const
+  template <typename T1, typename T2, typename T3>
+  SERAC_HOST_DEVICE auto operator()(const T1&, const T2&, const T3& temperature_gradient) const
   {
-    return -1.0 * conductivity_ * temperature_gradient;
+    return MaterialResponse{.density                = density_,
+                            .specific_heat_capacity = specific_heat_capacity_,
+                            .heat_flux              = -conductivity_ * temperature_gradient};
```

Comment:
I don't have a strong sense about the relative benefits/costs of switching to the MaterialResponse versus individual calls for each property. On the plus side, I see:
* provides consistency across different modeling physics. Heat conduction has three parameters. Structural modeling might have a different number. There is always one consistent operator() return a struct with the right number of material quantities.
* all parameters are treated equally one isn't treated with operator() and the others with individual methods
* potentially better performance because more computation per function call (guessing a very small effect if any)
Potential costs:
* another type to remember (albeit a single one)
* aggressive adoption of C++ 20 syntactic sugar for struct initialization (I think LiDO is at C++ 17)
* potential overhead of struct construction/return
* what are the missing arguments T1 and T2?

Reply:
I think the benefits outweigh the potential costs.

Reply:
So we are technically still on C++17, but our compilers (gcc and clang) both support designated initializers despite it not being required until C++20.

Reply:
>     + potentially better performance because more computation per function call (guessing a very small effect if any)

I mocked up a little example showing the assembly generated for doing a calculation with a struct that requires 3 member function calls to get separate quantities (like the "old" way), and an equivalent calculation that uses a struct that returns a "tuple" of those 3 values in one call (like the "new" way):

https://godbolt.org/z/xasvhcs4o

It appears that, when compiling with optimization, the generated assembly is practically identical.

>     - potential overhead of struct construction/return

C++17 actually introduced some language-level changes that make this sort of thing less expensive (or "free", in many cases). Previously, copy elision was a compiler optimization, so it was left up to the compiler to decide on a case-by-case basis when things would be elided or not. This meant that a lot of developers just assumed the worst case scenario (that return statements incur additional overhead) and avoided returning objects altogether in the name of "performance". But it seems like that's really not the case any more, there are a number of ways to get the clarity of a return value (instead of mixing inputs and outputs in the argument list) without the performance drawbacks!